core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala

```diff
@@ -0,0 +1,138 @@
+        val dao = new ByteArrayJournalDao(db, profile, journalConfig, SerializationExtension(system))
+
+        val payloadSize = 5000 // 5000 bytes
+        val eventsPerBatch = 1000
+
+        val numberOfInsertBatches = {
+          // calculate the number of batches using a factor to make sure we go a little bit over the limit
+          (maxMem.toInt / (payloadSize * eventsPerBatch) * 1.2).round.toInt
+        }
+        val totalMessages = numberOfInsertBatches * eventsPerBatch
+        val totalMessagePayload = totalMessages * payloadSize
+        log.info(
+          s"maxMem: $maxMem, batches: $numberOfInsertBatches (with $eventsPerBatch events), total messages: $totalMessages, total msgs size: $totalMessagePayload")
+
+        // payload can be the same when inserting to avoid unnecessary memory usage
+        val payload = Array.fill(payloadSize)('a'.toByte)
+
+        val lastInsert =
+          Source
+            .fromIterator(() => (1 to numberOfInsertBatches).toIterator)
+            .mapAsync(1) { i =>
+              val end = i * eventsPerBatch
+              val start = end - (eventsPerBatch - 1)
+              log.info(s"batch $i - events from $start to $end")
+              val atomicWrites =
+                (start to end).map { j =>
+                  AtomicWrite(immutable.Seq(PersistentRepr(payload, j, persistenceId)))
+                }.toSeq
+
+              dao.asyncWriteMessages(atomicWrites).map(_ => i)
+            }
+            .runWith(Sink.last)
+
+        // wait until we write all messages
+        // being very generous, 1 second per message
+        lastInsert.futureValue(Timeout(totalMessages.seconds))
+
+        log.info("Events written, starting replay")
+
+        val messagesSrc =
+          dao.messagesWithBatch(persistenceId, 0, totalMessages, batchSize = 100, None)
+        val doneFut =
+          messagesSrc.runForeach {
+            case Success(repr) =>
+              if (repr.sequenceNr % 100 == 0)
+                log.info(s"fetched: ${repr.persistenceId} - ${repr.sequenceNr}/${totalMessages}")
+            case Failure(exception) => println(exception)
+          }
+
+        // wait until we read all messages
+        // being very generous, 1 second per message
+        doneFut.futureValue(Timeout(totalMessages.seconds))
+      }
+    }
+  }
+}
+
+class PostgresJournalDaoStreamMessagesMemoryTest
+    extends JournalDaoStreamMessagesMemoryTest("postgres-application.conf")
+    with PostgresCleaner
+
+class MySQLJournalDaoStreamMessagesMemoryTest
+    extends JournalDaoStreamMessagesMemoryTest("mysql-application.conf")
+    with MysqlCleaner
+
+class OracleJournalDaoStreamMessagesMemoryTest
+    extends JournalDaoStreamMessagesMemoryTest("oracle-application.conf")
+    with OracleCleaner
+
+class SqlServerJournalDaoStreamMessagesMemoryTest
+    extends JournalDaoStreamMessagesMemoryTest("sqlserver-application.conf")
+    with SqlServerCleaner
```

Comment:
I ran this with `sbt -J-Xms96M -J-Xmx96M` and changed fork to false.

* Postgres and SqlServer are successful.
* Oracle and MySQL suffers from OutOfMemoryError already when **writing** the first batch

Reply:
With the latest test it works as expected for all databases.

Interesting is that it also works before the fix with Oracle and MySQL. Test failed before the fix with Postgres and SqlServer.

Reply:
Are you sure about this? When I tested it worked for Oracle and SqlServer before the fix. It seems that their driver don't fetch and keep everything in memory. 

I did observe the issue in Postgres and MySQL. Then, with the first impl (#351), it solved for Postgres, but not for MySQL.

Reply:
see my comments here: https://github.com/akka/akka-persistence-jdbc/pull/351#issuecomment-596651611

Oracle docker has some memory limitations and MySQL not working at all with fix in #351

Reply:
I ran this test one more time with
```
val messagesSrc = dao.messages(persistenceId, 0, totalMessages, totalMessages)
```

which corresponds to the old implementation.



* Postgress fail: Used heap before 13 MB, after 171 MB
* SqlServer success: Used heap before 14 MB, after 15 MB
* Oracle success: Used heap before 35 MB, after 15 MB
* MySQL fail: Used heap before 15 MB, after 95 MB

meaning that my previous comment here was wrong, I must have mixed up MySQL and SqlServer